llvm/lib/Target/DirectX/DXILShaderFlags.cpp

```diff
@@ -61,6 +66,21 @@ void ModuleShaderFlags::initialize(const Module &M) {
     CombinedSFMask.merge(CSF);
   }
   llvm::sort(FunctionFlags);
+  // Propagate shader flag mask of functions to their callers.
+  while (!WorkList.empty()) {
+    const Function *Func = WorkList.pop_back_val();
+    if (!Func->user_empty()) {
+      const ComputedShaderFlags &FuncSF = getFunctionFlags(Func);
+      // Update mask of callers with that of Func
+      for (const auto User : Func->users()) {
+        if (const CallInst *CI = dyn_cast<CallInst>(User)) {
+          const Function *Caller = CI->getParent()->getParent();
+          if (mergeFunctionShaderFlags(Caller, FuncSF))
+            WorkList.push_back(Caller);
```

Comment:
This approach is very inefficient. Consider a chain of functions, `f1`, `f2`, `f3`..., where `f1` has no flags, `f2` has one flag, `f3` has another flag, etc. In this case we might process `f1`, then add it to the worklist again while processing `f2`, process it again, and then add both `f2` to the worklist while processing `f3`, and then add `f1` to be processes *again* while processing `f2`, and so on.

The correct way to do this type of thing is to walk the call graph in post-order. One way to do this is by switching this pass to a [CallGraphSCCPass (old PM)](https://llvm.org/docs/WritingAnLLVMPass.html#the-callgraphsccpass-class`) / [CGSCCPass (new PM)](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/CGSCCPassManager.h#L136).

Reply:
Ah, since this is an analysis and it's used by module passes we won't be able to simply convert to an SCC pass. As is, we may need to construct the call graph manually to do this in the more efficient way. It's probably fine to just construct a [CallGraph](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/CallGraph.h#L66) and use [SCCIterator](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/ADT/SCCIterator.h#L40) to process the functions in post order.

Reply:
Changes made to build a `CallGraph` and use `scc_iterator` to create and propagate shader flags masks for module functions.